torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+                    )
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_sig]
+                    )
+                else:
+                    raise RuntimeError(
+                        f"non compute op not supporting {arg_node_type}! "
+                    )
+
+                # finished processing this non-compute node
+                continue
+
+            # for computatation nodes, we need to check all the inputs
+            input_args = node.all_input_nodes
+            input_node_types = pytree.tree_map(
+                lambda x: dp_strategy_map[x].node_type, input_args
+            )
+            input_specs = []
+            if node in dp_strategy_map:
+                # found a param_grad node that already have output pre-filled spec
+                # fill in the expected input specs for the pre-filled strategy
+                node_type = dp_strategy_map[node].node_type
+                assert node_type == NodeType.GRAD
+                produce_param_grad_strat = dp_strategy_map[node].strategies
+                has_activation = False
+                for arg in input_args:
+                    arg_node_type = dp_strategy_map[arg].node_type
+                    if arg_node_type == NodeType.ACT:
+                        # activation sharded
+                        has_activation = True
+                        act_spec, _ = batch_dim_analzer.get_batch_dim_shard_spec(
+                            arg, mesh
+                        )
+
+                        input_specs.append(act_spec)
+
+                if has_activation:
+                    assert len(produce_param_grad_strat) == 1
+                    produce_param_grad_strat[0].input_specs = input_specs
+            else:
+                if NodeType.ACT in input_node_types:
+                    # param + activation, build up acceptable strategy
+                    # param must be replicated, activation must be sharded
+                    for arg in input_args:
+                        arg_strategies = dp_strategy_map[arg]
+                        node_type = arg_strategies.node_type
+                        if node_type == NodeType.ACT:
+                            # activation must stay sharded
+                            act_spec, _ = batch_dim_analzer.get_batch_dim_shard_spec(
+                                arg, mesh
+                            )
+
+                            input_specs.append(act_spec)
+                        elif node_type == NodeType.PARAM:
+                            # param must be replicated
+                            input_specs.append(
+                                DTensorSpec(mesh=mesh, placements=[Replicate()])
+                            )
+                        else:
+                            raise RuntimeError(
+                                f"Expecting node with parameter and activation, but found {input_node_types}! "
```

Comment:
(can be done in follow up PRs): ops like `nll_loss` does take two activation tensors as input?